packages/react-urql/src/hooks/state.ts

```diff
@@ -9,10 +9,21 @@ export const initialState = {
   operation: undefined,
 };
 
-const isShallowDifferent = (a: any, b: any) => {
-  if (typeof a != 'object' || typeof b != 'object') return a !== b;
-  for (const x in a) if (!(x in b)) return true;
-  for (const x in b) if (a[x] !== b[x]) return true;
+/**
+ * Checks if two objects are shallowly different with a special case for
+ * 'operation' where it compares the key if they are not the otherwise equal
+ */
+const isShallowDifferent = <T extends Record<string, any>>(a: T, b: T) => {
+  for (const key in a) if (!(key in b)) return true;
+  for (const key in b) {
+    if (a[key] !== b[key]) {
+      // Two operations are considered equal if they have the same key
+      if (key === 'operation' && a.operation?.key === b.operation?.key) {
+        continue;
+      }
+      return true;
+    }
+  }
```

Comment:
Ha, nice catch 😅 I didn't expect this old code to have such a severe lingering issue.
Re. the implementation, we currently avoid nullish chaining and coalescing operators. We don't have a lint set up for that right now, so obviously you couldn't have known, but it's basically to avoid some issues with quirky JS engines that are still in range of our support. Currently, this will lead to some excessively large polyfill code

We could just avoid them by type casting but we can also do

```suggestion
const isShallowDifferent = <T extends Record<string, any>>(a: T, b: T) => {
  for (const key in a) if (!(key in b)) return true;
  for (const key in b) {
    if (
      (key === 'operation' && a[key].key === b[key].key) ||
      a[key] !== b[key]
    ) return true;
  }
```

...since we don't have exact types here anyway

Reply:
If `a[key].key === b[key].key` then we don't want to return true, I can adjust it to not use the operator

Reply:
I introduced a helper function to keep it readable but if you'd prefer not to have an extra function call for performance reasons then I can inline it